packages/truffle-core/lib/testing/testsource.js

```diff
@@ -13,87 +13,83 @@ TestSource.prototype.require = function() {
 };
 
 TestSource.prototype.resolve = function(import_path, callback) {
-  var self = this;
+  const self = this;
 
   if (import_path === "truffle/DeployedAddresses.sol") {
     return find_contracts(this.config.contracts_directory, function(
       err,
-      source_files
+      sourceFiles
     ) {
       // Ignore this error. Continue on.
 
-      fs.readdir(self.config.contracts_build_directory, function(
-        err,
-        abstraction_files
-      ) {
-        if (err) return callback(err);
-
-        var mapping = {};
+      let abstraction_files;
+      try {
+        abstraction_files = fs.readdirSync(
+          self.config.contracts_build_directory
+        );
+      } catch (error) {
+        return callback(error);
+      }
+
+      const mapping = {};
+
+      const blacklist = ["Assert", "DeployedAddresses"];
+
+      // Ensure we have a mapping for source files and abstraction files
+      // to prevent any compile errors in tests.
+      sourceFiles.forEach(file => {
+        const name = path.basename(file, ".sol");
+        if (blacklist.indexOf(name) >= 0) return;
+        mapping[name] = false;
+      });
 
-        var blacklist = ["Assert", "DeployedAddresses"];
+      abstraction_files.forEach(file => {
+        const name = path.basename(file, ".json");
+        if (blacklist.indexOf(name) >= 0) return;
+        mapping[name] = false;
+      });
 
-        // Ensure we have a mapping for source files and abstraction files
-        // to prevent any compile errors in tests.
-        source_files.forEach(function(file) {
-          var name = path.basename(file, ".sol");
-          if (blacklist.indexOf(name) >= 0) return;
-          mapping[name] = false;
+      const promises = abstraction_files.map(file => {
+        return new Promise((accept, reject) => {
+          fs.readFile(
+            path.join(self.config.contracts_build_directory, file),
+            "utf8",
+            function(err, body) {
+              if (err) return reject(err);
+              accept(body);
+            }
+          );
```

Comment:
This seems a little clunky -- do you need the `return new Promise` part?  Since `fs.readFile` is async won't it already return a promise? Feel free to ignore me if I'm missing something here! 

You could also just wrap the map here in a `Promise.all` and chain the `then` code from below to it since line 65 seems to be the only place you use the `promises` variable, but that's just a stylistic thing, I think its fine either way.

Reply:
I don't think those fs methods return promises but rather have a callback interface...HOWEVER, it looks like we can just drop in `fs-extra` instead. Those asynchronous methods DO return Promises. That way it will be more prettier ;)